public/app/features/dashboard-scene/utils/dashboardSessionState.ts

```diff
@@ -0,0 +1,115 @@
+import { UrlQueryMap, urlUtil } from '@grafana/data';
+import { config, locationService } from '@grafana/runtime';
+
+import { DashboardScene } from '../scene/DashboardScene';
+
+export const PRESERVED_SCENE_STATE_KEY = `grafana.dashboard.preservedUrlFiltersState`;
+
+export function restoreDashboardStateFromLocalStorage(dashboard: DashboardScene) {
+  const preservedUrlState = window.sessionStorage.getItem(PRESERVED_SCENE_STATE_KEY);
+
+  if (preservedUrlState) {
+    const preservedQueryParams = new URLSearchParams(preservedUrlState);
+    const currentQueryParams = locationService.getSearch();
+
+    // iterate over preserved query params and append them to current query params
+    preservedQueryParams.forEach((value, key) => {
+      currentQueryParams.append(key, value);
+    });
+
+    // remove duplicate query params
+    const deduplicatedQueryParams = deduplicateQueryParams(currentQueryParams);
+
+    for (const key of Array.from(deduplicatedQueryParams.keys())) {
+      // preserve non-variable query params, i.e. time range
+      if (!key.startsWith('var-')) {
+        continue;
+      }
+
+      // remove params for variables that are not present on the target dashboard
+      if (!dashboard.state.$variables?.getByName(key.replace('var-', ''))) {
+        deduplicatedQueryParams.delete(key);
+      }
+    }
+
+    const finalParams = deduplicatedQueryParams.toString();
+    if (finalParams) {
+      locationService.replace({
+        search: finalParams,
+      });
+    }
+  }
+}
+
+/**
+ * Scenes behavior that will capture currently selected variables and time range and save them to local storage, so that they can be applied when the next dashboard is loaded.
+ */
+export function preserveDashboardSceneStateInLocalStorage(scene: DashboardScene) {
+  if (!config.featureToggles.preserveDashboardStateWhenNavigating) {
+    return;
+  }
+
+  return () => {
+    // Skipping saving state for default home dashboard
+    if (!scene.state.uid) {
+      return;
+    }
+    const variables = scene.state.$variables?.state.variables;
+    const timeRange = scene.state.$timeRange;
+
+    let urlStates: UrlQueryMap = variables
```

Comment:
Still not sure why we are not using UrlSyncManager.getUrlState(scene)  here

Reply:
IMO it's better to opt-in for params to sync, rather than removing items from the full state (may include setting, panel edit, view panel id etc...).

Reply:
@dprokop but instead of removing state you don't care about just include from & to and all that start with `var-` , that is basically the same as the code here but much smaller and less complex

Reply:
just pushed 👍